src/lxc/network.c

```diff
@@ -2818,47 +2856,108 @@ static int lxc_setup_l2proxy(struct lxc_netdev *netdev) {
+
+		/* IPVLAN requires a route to local-loopback to trigger l2proxy. */
+		if (netdev->type == LXC_NET_IPVLAN) {
+			err = lxc_ipv6_dest_add(lo_ifindex, &inet6dev->addr, 128);
+			if (err < 0) {
+				ERROR("Failed to add ipv6 dest \"%s\" for network device \"%s\"", bufinet6, loDev);
+				return minus_one_set_errno(-err);
+			}
+		}
 	}
 
 	return 0;
 }
 
+static int lxc_delete_ipv4_l2proxy(struct in_addr *ip, char *link, unsigned int lo_ifindex) {
+	char bufinet4[INET_ADDRSTRLEN];
+	unsigned int errCount = 0;
+
+	if (!inet_ntop(AF_INET, ip, bufinet4, sizeof(bufinet4))) {
+		SYSERROR("Failed to convert IP for l2proxy ipv4 removal on dev \"%s\"", link);
+		return minus_one_set_errno(EINVAL);
+	}
+
+	/* If a local-loopback ifindex supplied remove the static route to the lo device. */
+	if (lo_ifindex > 0) {
+		if (lxc_ipv4_dest_del(lo_ifindex, ip, 32) < 0) {
+			errCount++;
+			ERROR("Failed to delete ipv4 dest \"%s\" for network ifindex \"%u\"", bufinet4, lo_ifindex);
+		}
+	}
+
+	/* If link is supplied remove the IP neigh proxy entry for this IP on the device. */
+	if(link[0] != '\0') {
+		if (lxc_del_ip_neigh_proxy(bufinet4, link) < 0)
+			errCount++;
+	}
+
+	if (errCount > 0)
+		return minus_one_set_errno(EINVAL);
+
+	return 0;
+}
+
+static int lxc_delete_ipv6_l2proxy(struct in6_addr *ip, char *link, unsigned int lo_ifindex) {
+	char bufinet6[INET6_ADDRSTRLEN];
+	unsigned int errCount = 0;
+
+	if (!inet_ntop(AF_INET6, ip, bufinet6, sizeof(bufinet6))) {
+		SYSERROR("Failed to convert IP for l2proxy ipv6 removal on dev \"%s\"", link);
+		return minus_one_set_errno(EINVAL);
+	}
+
+	/* If a local-loopback ifindex supplied remove the static route to the lo device. */
+	if (lo_ifindex > 0) {
+		if (lxc_ipv6_dest_del(lo_ifindex, ip, 128) < 0) {
+			errCount++;
+			ERROR("Failed to delete ipv6 dest \"%s\" for network ifindex \"%u\"", bufinet6, lo_ifindex);
+		}
+	}
+
+	/* If link is supplied remove the IP neigh proxy entry for this IP on the device. */
+	if(link[0] != '\0') {
+		if (lxc_del_ip_neigh_proxy(bufinet6, link) < 0)
+			errCount++;
+	}
+
+	if (errCount > 0)
+		return minus_one_set_errno(EINVAL);
+
+	return 0;
+}
+
 static int lxc_delete_l2proxy(struct lxc_netdev *netdev) {
+	unsigned int lo_ifindex = 0;
+	unsigned int errCount = 0;
 	struct lxc_list *cur, *next;
 	struct lxc_inetdev *inet4dev;
 	struct lxc_inet6dev *inet6dev;
-	char bufinet4[INET_ADDRSTRLEN], bufinet6[INET6_ADDRSTRLEN];
-	int err = 0;
 
-	lxc_list_for_each_safe(cur, &netdev->ipv4, next) {
-		inet4dev = cur->elem;
-		if (!inet_ntop(AF_INET, &inet4dev->addr, bufinet4, sizeof(bufinet4))) {
-			err = -1;
-			SYSERROR("Failed to convert IP for l2proxy removal on dev \"%s\"", netdev->link);
-			continue; /* Try to remove any other l2proxy entries */
+	/* Perform IPVLAN specific checks. */
+	if (netdev->type == LXC_NET_IPVLAN) {
+		/* Retrieve local-loopback interface index for use with IPVLAN static routes. */
+		lo_ifindex = if_nametoindex(loDev);
+		if (lo_ifindex == 0) {
```

Comment:
Hm, above you use `if (!)` and here `if (== 0)`. Can you please make it consistent? :)